File: packages/core-js/src/roundtrip.ts
/**
 * Parse a Markdown table back to JSON (array of objects).
 * Best-effort: works for flat tables produced by jsonmd.
 */
export function mdToJson(md: string): unknown {
  const lines = md.split('\n');
  const tables: { headers: string[]; rows: Record<string, unknown>[] }[] = [];

  let i = 0;
  while (i < lines.length) {
    const line = lines[i].trim();

    // Detect a table: line starts with |
    if (line.startsWith('|') && i + 1 < lines.length) {
      const headers = parseCells(line);
      const nextLine = lines[i + 1]?.trim();

      // Verify separator row
      if (nextLine && nextLine.startsWith('|') && nextLine.includes('---')) {
        i += 2; // skip header + separator
        const rows: Record<string, unknown>[] = [];

        while (i < lines.length && lines[i].trim().startsWith('|')) {
          const cells = parseCells(lines[i].trim());
          const row: Record<string, unknown> = {};
          for (let j = 0; j < headers.length; j++) {
            row[headers[j]] = coerce(cells[j] ?? '');
          }
          rows.push(row);
          i++;
        }

        // Key-Value table detection
        if (headers.length === 2 && headers[0] === 'Key' && headers[1] === 'Value') {
          const obj: Record<string, unknown> = {};
          for (const row of rows) {
            const key = String(row['Key'] ?? '');
            if (key) obj[key] = row['Value'];
          }
          tables.push({ headers, rows: [obj as Record<string, unknown>] });
        } else {
          tables.push({ headers, rows });
        }
        continue;
      }
    }

    i++;
  }

  if (tables.length === 0) return null;
  if (tables.length === 1) {
    const t = tables[0];
    // Key-Value table
    if (t.headers.length === 2 && t.headers[0] === 'Key' && t.headers[1] === 'Value') {
      return t.rows[0];
    }
    return t.rows;
  }

  // Multiple tables - return array of all rows
  return tables.flatMap((t) => t.rows);
}

function parseCells(line: string): string[] {
  // Split by | but not \|
  const inner = line.replace(/^\|/, '').replace(/\|$/, '');
  const cells: string[] = [];
  let current = '';
  for (let i = 0; i < inner.length; i++) {
    if (inner[i] === '\\' && i + 1 < inner.length && inner[i + 1] === '|') {
      current += '|';
      i++;
    } else if (inner[i] === '|') {
      cells.push(current.trim());
      current = '';
    } else {
      current += inner[i];
    }
  }
  cells.push(current.trim());
  return cells;
}

function coerce(value: string): unknown {
  if (value === '') return null;
  if (value === 'true') return true;
  if (value === 'false') return false;
  if (value === 'null') return null;

  // Try number
  const num = Number(value);
  if (!isNaN(num) && value.trim() !== '') {
    return num;
  }

  return value;
}

File: packages/core-python/jsonmd/roundtrip.py
"""Parse Markdown tables back to JSON."""

from __future__ import annotations

from typing import Any


def md_to_json(md: str) -> Any:
    """Parse a Markdown table back to JSON (array of objects). Best-effort."""
    lines = md.split("\n")
    tables: list[dict] = []

    i = 0
    while i < len(lines):
        line = lines[i].strip()

        if line.startswith("|") and i + 1 < len(lines):
            headers = _parse_cells(line)
            next_line = lines[i + 1].strip() if i + 1 < len(lines) else ""

            if next_line.startswith("|") and "---" in next_line:
                i += 2
                rows = []

                while i < len(lines) and lines[i].strip().startswith("|"):
                    cells = _parse_cells(lines[i].strip())
                    row = {}
                    for j, header in enumerate(headers):
                        row[header] = _coerce(cells[j] if j < len(cells) else "")
                    rows.append(row)
                    i += 1

                # Key-Value table detection
                if len(headers) == 2 and headers[0] == "Key" and headers[1] == "Value":
                    obj = {}
                    for row in rows:
                        key = str(row.get("Key", ""))
                        if key:
                            obj[key] = row["Value"]
                    tables.append({"headers": headers, "rows": [obj]})
                else:
                    tables.append({"headers": headers, "rows": rows})
                continue

        i += 1

    if not tables:
        return None
    if len(tables) == 1:
        t = tables[0]
        if len(t["headers"]) == 2 and t["headers"][0] == "Key" and t["headers"][1] == "Value":
            return t["rows"][0]
        return t["rows"]

    result = []
    for t in tables:
        result.extend(t["rows"])
    return result


def _parse_cells(line: str) -> list[str]:
    """Split a table row by | but not \\|."""
    inner = line.strip("|")
    cells = []
    current = ""
    i = 0
    while i < len(inner):
        if inner[i] == "\\" and i + 1 < len(inner) and inner[i + 1] == "|":
            current += "|"
            i += 2
        elif inner[i] == "|":
            cells.append(current.strip())
            current = ""
            i += 1
        else:
            current += inner[i]
            i += 1
    cells.append(current.strip())
    return cells


def _coerce(value: str) -> Any:
    """Coerce a string value to an appropriate Python type."""
    if value == "":
        return None
    if value == "true":
        return True
    if value == "false":
        return False
    if value == "null":
        return None

    try:
        # Try int first
        if "." not in value and "e" not in value.lower():
            return int(value)
        return float(value)
    except (ValueError, OverflowError):
        pass

    return value
